containers: link image registries with their openshift services

moolitayer opened the pull request:
container service view: added registry

![manageiq services](https://cloud.githubusercontent.com/assets/3010449/11868942/4109b1ae-a4c4-11e5-9bfb-fdadab6a6546.png)

container image registry view: added service and pods

![manageiq image registries](https://cloud.githubusercontent.com/assets/3010449/11955355/09afcc26-a8ba-11e5-813b-bee4f6d19291.png)

container image registry view: list of services

![manageiq image registries s list](https://cloud.githubusercontent.com/assets/3010449/11955373/29f6a298-a8ba-11e5-8435-d8b4803a7d5e.png)


container image registry view: list of pods

![screenshot from 2015-12-10 21-55-40](https://cloud.githubusercontent.com/assets/3010449/11726842/4ec19202-9f89-11e5-9c50-0f16564e7ebb.png)

moolitayer commented:
@simon3z @zeari please review

zeari commented on app/models/container_image_registry.rb:
```diff
@@ -4,6 +4,8 @@ class ContainerImageRegistry < ActiveRecord::Base
   has_many :container_images, :dependent => :destroy # What about deleted registry but containers are still running
   has_many :containers, :through => :container_images
+  has_one :container_service
+  has_many :container_groups, :through => :container_service
```
@moolitayer Why the relation to ```container_groups```?

moolitayer replied:
this was in the requirement and I display them as well.
As I see it if there is a service that creates a registry but no pods
it's important for the user to see since it will explain why the registry is actually down.

zeari commented on app/models/manageiq/providers/kubernetes/container_manager/refresh_parser.rb:
```diff
@@ -212,7 +212,16 @@ def parse_service(service)
+          :container_image_registry, :by_host_and_port, "#{new_result[:portal_ip]}:#{pc[:port]}"
+        )
+        if os_registry
+          puts "found a match os_registry[:name]: #{os_registry[:name]} new_result[:name]:#{new_result[:name]}"
```
@moolitayer Intentional or left over from debugging?

moolitayer replied:
leftover, thanks. Rubocop is also angry

zeari commented:
I think this needs tests. Other than that, this is pretty much what I expected.

moolitayer commented:
@miq-bot add_label providers/containers, ui

moolitayer commented:
added tests

simon3z commented:
@moolitayer anything to do about the navbar?

jrafanie commented:
@moolitayer this was conflicted by #5841, I believe, ping @chrisarcand or I if you need help

moolitayer commented:
> @moolitayer this was conflicted by #5841, I believe, ping @chrisarcand or I if you need help

thanks

added navbar, improved testing, fixed error on service screens with no image registries and anther refresh error

@zeari @simon3z please review, thanks!

zeari commented on app/models/container_service.rb:
```diff
@@ -12,6 +12,7 @@ class ContainerService < ActiveRecord::Base
   has_many :labels, -> { where(:section => "labels") }, :class_name => "CustomAttribute", :as => :resource, :dependent => :destroy
   has_many :selector_parts, -> { where(:section => "selectors") }, :class_name => "CustomAttribute", :as => :resource, :dependent => :destroy
   has_many :container_nodes, -> { distinct }, :through => :container_groups
+  belongs_to :container_image_registry
```
Just wondering how openshift works exactly: Theres a service that caters to an image registry? 
@moolitayer @abonas

moolitayer replied:
> Just wondering how openshift works exactly: Theres a service that caters to an image registry? 
> @moolitayer @abonas

[integrated openshift registry](https://docs.openshift.org/latest/architecture/infrastructure_components/image_registry.html#integrated-openshift-registry)

@zeari @simon3z please review, thanks!

abonas replied:
@moolitayer perhaps the same logic should be added for routes

moolitayer commented:
@zeari @simon3z please review

zeari commented on app/helpers/container_summary_helper.rb:
```diff
@@ -82,8 +82,7 @@ def textual_guest_applications
     object = @record.container_image_registry
-
-    if object.nil?
+    if object.nil? && @record.class == ContainerImage
```
Why did you need this?

moolitayer replied:
We talked about this actually. This function previously assumed only the container image screen called a container image registry relation. This cased an exception for container service with no container image registry attached.

simon3z replied:
@moolitayer this should be at least `kind_of?`... although even that is not optimal.
Best thing is to handle that with `respond_to?`. cc @zeari

zeari commented:
Apart from conflicts and a comment I left, 
LGTM :+1:

moolitayer commented:
rebased

simon3z commented:
@moolitayer technically a single `image_registry` could be referenced by multiple `container_services`. The backend (db, entities) seems to support it already, probably you just need to handle that in the UI.

moolitayer commented:
> @moolitayer technically a single image_registry could be referenced by multiple container_services. The backend (db, entities) seems to support it already, probably you just need to handle that in the UI.

how is that possible? the services define the registry

simon3z commented:
> how is that possible? the services define the registry

Look at the service... what is preventing you from creating another service (different name) but same/similar selector for the registry pod.

abonas commented:
> how is that possible? the services define the registry

@moolitayer the registry is running as a pod, not a service. service is passing traffic to it.

moolitayer commented:
update relation definition, tests and screen.
updated screenshots

simon3z commented:
@moolitayer any basic spec test to add?

abonas commented on spec/models/manageiq/providers/openshift/container_manager/refresher_spec.rb:
```diff
@@ -95,7 +96,7 @@ def assert_specific_container_node
   end
 
-  def assert_specific_container_service
+  def assert_specific_container_services
```
why is the method renamed? (while there aren't other changes in the method body)

moolitayer replied:
it now looks at more then one service in line 111

blomquisg commented:
@simon3z any feedback on this one?  Looks ready, but wasn't sure.

moolitayer commented:
@simon3z @zeari please review

zeari commented:
LGTM on my part :+1:

simon3z commented:
@zeari @moolitayer you may have missed my comment on `respond_to?`. Please address that first.

moolitayer commented:
rebased, updated container_summary_helper to use kind_of? (since there is no respond_to? X to use here)
@simon3z please review

simon3z commented on app/helpers/container_summary_helper.rb:
```diff
@@ -94,8 +94,7 @@ def textual_guest_applications
+    if object.nil? && @record.kind_of?(ContainerImage)
       {
         :label => ui_lookup(:model => ContainerImageRegistry.name),
         :image => "container_image_registry_unknown",
```
Is the string returned by `display_registry` localized here? (cc @zeari )

zeari replied:
Youre right, Ill fix it in a separate patch.

simon3z commented:
> rebased, updated container_summary_helper to use kind_of? (since there is no respond_to? X to use here)

It seems that you use it to check if `display_registry` is available, so your `respond_to?` should be on that.

moolitayer commented:
> rebased, updated container_summary_helper to use kind_of? (since there is no respond_to? X to use here)
> It seems that you use it to check if display_registry is available, so your respond_to? should be on that.

right, I missed that some how. Updated. @simon3z please review

simon3z commented:
LGTM :+1: cc @blomquisg @chessbyte
